### ISSUE TITLE
As a user, I want to upload a pair of files via an easy interface

### ISSUE THREAD

**Issue by TangoYankee** [openoakland/woeip#31]
**Is your feature request related to a problem? Please describe.**
The current interface depends on clicking on a button, which opens up file explorer. I would like to already have my location open and then drag the files onto the page. 

**Describe the solution you'd like**
1) Upon navigating to the `add data`/`upload` page, I should see an area where I can drag and drop files.
2) In its initial state, the box should let me know that I can drag my gps and dustrak files into the drop box
3) When dragging the files, if I hover for a few seconds, the page should instruct me to let go of the mouse to drop the files
4) As the files are uploaded, I should receive feedback that the files are being processed
5) If there is an error in the upload process, I should receive feedback for the following cases:
    - The incorrect number of files were uploaded (too many or too few)
    - The incorrect combination of files were uploaded (not one csv and one gps file)
    - General error message
6) There will still be a button that can optionally be used to upload the files
7) The upload process should be triggered automatically- once two files of the correct type are identified

**Describe alternatives you've considered**
We are replacing the current style where the upload button is the only option

**Additional context**
1) In the wireframes, there is a `need help` button. This button will be saved for later iterations.
2) There is an additional error state, in which the dates of the gps and dustrak files don't match. As this demands parsing through the files, it will be reserved for another issue.

**Labels**
- [x] Add a status label
  - `backlog`
  - `discussing`
  - `ready`
  - `implementing`
- [x] Add additional labels as needed

### For Developers
**Affected Components**
The entire `upload` page will require a `view`, `form`, and `template`. However, this issue should be limited to the user interface build in javascript, on top of the other parts. For this issue, the affected components are:
- `upload.html`
- `app.js`

**Technical Resources**
This issue is dependent on issue # , which establishes the `view`, `form`, `url`, and `template` on which the interface is built. 

Possible framework: [jQuery-File-Upload](https://github.com/blueimp/jQuery-File-Upload)

Precedents and Tutorials:
[Mozilla Developer Resource](https://developer.mozilla.org/en-US/docs/Web/API/HTML_Drag_and_Drop_API/File_drag_and_drop)
[Multiple File Upload tutorial](https://simpleisbetterthancomplex.com/tutorial/2016/11/22/django-multiple-file-upload-using-ajax.html)
[Multiple FIles Tutorial Source Code](https://github.com/sibtc/multiple-file-upload)
[Django Files Tutorial](https://simpleisbetterthancomplex.com/tutorial/2016/08/01/how-to-upload-files-with-django.html)
[JQuery File Upload Sample](https://github.com/blueimp/jQuery-File-Upload/blob/master/basic-plus.html)

**Design Documentation**
[Wireframes V3](https://github.com/openoakland/woeip/files/2959738/WOEIP-UploadFlow-v3.pdf)

**Comment by Ethan-bradley** [comment]
Here is a basic html file (have to convert from txt to html) of what the upload page would look like: 
[test3.txt](https://github.com/openoakland/woeip/files/2957841/test3.txt)

**Comment by TangoYankee** [comment]
@Ethan-bradley This functionality manipulates objects that are already in the web browser Can it be generalized to hold files that are dragged from a file explorer to the web browser?

**Comment by Ethan-bradley** [comment]
Yes, it is supposed to do that and seems to work on my computer.

**Comment by Ethan-bradley** [comment]
Here is the finished version of the UI interface (convert to html and js): 
[dragdrop.txt](https://github.com/openoakland/woeip/files/2959745/dragdrop.txt)
[dragdropjs.txt](https://github.com/openoakland/woeip/files/2959746/dragdropjs.txt)

**Comment by TangoYankee** [comment]
[Version Three of the Wireframes](https://github.com/openoakland/woeip/files/2959738/WOEIP-UploadFlow-v3.pdf) has some intermediate functionality dependent on JS. @theecrit would you be open to reducing the number of dynamic elements, to simplify the MVP?
1. Eliminate the `Hover state`
2. Change Step 1 to read `Drag and Drop Dustrak and GPS files here`
3. Eliminate `file matching state`

@Ethan-bradley I looked at the html. Could you bring the design of the steps more in line with the Wireframes and any accepted changes?

**Comment by Ethan-bradley** [comment]
Sure, no problem

**Comment by theecrit** [comment]
Yes on #1. Can you explain your reasoning on #2 & 3 so I have some context? Thanks.

**Comment by TangoYankee** [comment]
1. After thinking it over, we could actually keep this. It's not too complex.
2. My lazy attempt to make up for initially deleting the `hover` state. I'm not married to it.
3. We are using two languages: Python and Javascript. Each language is capable of doing the requested function. However, we generally uses Javascript for Design only. We keep logic and data management in Python/Django. By making `file matching` something that is basically live updated to the user, we are pushing logic and data management to Javascript. I have full faith that we could eventually support this capability. However, it would modify the current delegation of responsibilities. This could cause delays in deployment, as we build the logistical capability to implement it. <br>
Returning to the original delegation of language responsibilities means eliminating the `live update`. The website would instead start the 'upload' process once a user pressed `submit`. The data would be sent from the front end (Javascript) to the back end (Python). The backend would process the data and send it back to the front end to display. At this point, the front end would display either the `success` or the `error` state, based on what the back end returned.

**Comment by theecrit** [comment]
1. Cool, let's keep it in for now. We can de-scope if it becomes a time-suck.
2. n/a
3. Got it; makes sense to me. Let's de-scope from MVP. I'll update wires this week.

**Comment by TangoYankee** [comment]
@Ethan-bradley, you are the one working on the code of this issue. What are your thoughts on the  changes?

**Comment by Ethan-bradley** [comment]
@TangoYankee I agree on the second and third changes, as without a submit button the user may get confused on whether the files have been submitted (However, Jess may think that it adds more complexity). Additionally, for the second change it adds more clarity about what to do. Currently, the code already has the hover state, automatically updates, and automatically submits after dropping files (though submission through django forms has not yet been implemented). Additionally, I will look at deciding which file is which in the python code.

**Comment by Ethan-bradley** [comment]
Here is an updated version of the drag and drop:
[dragdrop.txt](https://github.com/openoakland/woeip/files/2986330/dragdrop.txt)
[dragdropjs.txt](https://github.com/openoakland/woeip/files/2986331/dragdropjs.txt)

**Comment by r-b-g-b** [comment]
Agreed with what you all have come up with in the discussion so far. 1) Keep hover, it's nice and doesn't add too much complexity, 2) move all logic to Django. The drag and drop can really be as simple as uploading two files to the server. The server can handle figuring out which is air Q and which is GPS, if they are properly formatted, etc. I'd suggest moving this to PR so we can go over the code itself.

**Comment by r-b-g-b** [comment]
Caveat about relegating logic to python: A few things should be handled on the front end prior to upload:
 - Set max file size that will upload
 - Limit file extensions that will upload

**Comment by TangoYankee** [comment]
@r-b-g-b I agree that it's ready for a prototype. @Ethan-bradley, could you push a branch to the repo? However, **don't open a pull request**, yet. A pull request should be just that- a request to be pulled into the code. For now, the team can look over the code in the shared branch. However, we should continue to have discussions about it in this Issue Thread.

**Comment by Ethan-bradley** [comment]
I'll push a branch to the repo tommorrow once the script is in place inside the django app.

**Comment by Ethan-bradley** [comment]
Here is the dragdrop interface branch:
https://github.com/openoakland/woeip/tree/eb/dragdrop2
The files are not uploaded yet to the app.

**Comment by TangoYankee** [comment]
@r-b-g-b What are you thinking for max file size?

**Comment by TangoYankee** [comment]
I had a thought about timezone. If we are saving the data to the database before allowing the user to confirm the data, then the dates will have already been converted from local to UTC time in the database. If the user changes the local time, then we will need to go through and change the time of every data point in the database. 

My recommendation is that we make the timezone a field on the original upload form. While we're at it, might as well include the `collected_by` field too.

**Comment by r-b-g-b** [comment]
Mmmmm largest file of the ~10 sessions I have is 2MB, so maybe like 25MB?

**Comment by Ethan-bradley** [comment]
Sounds good, I'll add that max file size in.

**Comment by nbilenko** [comment]
Django could handle the max upload size issue - here's a snippet with an example https://www.djangosnippets.org/snippets/1303/

**Comment by Ethan-bradley** [comment]
Thanks, I am currently using the javascript max file size function but I will look at the Django implementation as well.

**Comment by TangoYankee** [comment]
This PR looked like it was getting kinda big. I made #81 to handle the backend logic. That way, this PR can focus on the front end UI. My suggestion is to discuss #81, settle on a backend design, and then implement. From there, we can add on the front end from this issue. If ya'll disagree, let me know.

**Comment by Ethan-bradley** [comment]
@TangoYankee Currently I am planning to use ajax to send the data to Django and forms to handle the backend logic (currently none for the moment). Additionally, the javascript determines file type and maxfilesize before sending it to the views.py.

**Comment by Ethan-bradley** [comment]
@TangoYankee Actually, now the plan is to use a django form to send the information. Currently this is working, though the backend form needs to be edited.

**Comment by TangoYankee** [comment]
> @TangoYankee Actually, now the plan is to use a django form to send the information. Currently this is working, though the backend form needs to be edited.

Thank you for letting me know. Is there anything you need help with?

**Comment by theecrit** [comment]
Closing as "won't fix": since tech stack moved front-end to React, this issue is now covered by #129.
